### PR TITLE
Kubeadm: Change the marshal code to use ApiMachinery code.

### LINKED ISSUE #453: Stop using `yaml.Marshal`. Use the real API machinery code instead.
Do not do this: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/staticpod/utils.go#L131

as that has various side-effects and is not supported by the API Machinery team. For instance, `TypeMeta` is not populated correctly.

Instead, do this: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/apiclient/dryrunclient.go#L44-L53

### PR DESCRIPTION
**Pull request by medinatiger**
**What this PR does / why we need it**:
The PR change the k8s obj marshaling to use ApiMachinery code instead of plain yaml.Marshal which is known to have some side-effect.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
fixes kubernetes/kubeadm#453
 
**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go [hunk 1]
```diff
@@ -23,8 +23,9 @@ import (
 	"os"
 	"testing"
 
-	"github.com/ghodss/yaml"
+	"k8s.io/api/extensions/v1beta1"
```
Comment (luxas) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go]: everything here has now moved to using `apps/v1beta2` as you rebase

File: cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go [hunk 2]
```diff
@@ -102,7 +103,9 @@ spec:
 status: {}
 `
 
-	testAPIServerDaemonSet = `metadata:
+	testAPIServerDaemonSet = `apiVersion: extensions/v1beta1
```
Comment (luxas) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go]: apps/v1beta2 everywhere here

File: cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go (hunk 1 above)
Comment (luxas) [reply cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go]: we often import this as the group name; like `extensions` or `apps` respectively

File: cmd/kubeadm/app/util/marshal.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
Comment (luxas) [inline cmd/kubeadm/app/util/marshal.go]: now 2017 ;)

File: cmd/kubeadm/app/util/marshal.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	clientsetscheme "k8s.io/client-go/kubernetes/scheme"
+)
+
+// Marshal object into yaml.
+func MarshalToYaml(obj runtime.Object, gv schema.GroupVersion) ([]byte, error) {
```
Comment (luxas) [inline cmd/kubeadm/app/util/marshal.go]: we're enforcing golint now; you must have a valid comment

Comment (luxas) [reply cmd/kubeadm/app/util/marshal.go]: cc @sttts Should/can we move this somewhere else?

**Review by luxas — COMMENTED** [review]
Cool!
A few things have changed at HEAD now since you made this, so you have to rebase and then will have to update a few things.

Also we should talk to @kubernetes/sig-api-machinery-pr-reviews whether this is something that should exist in a more generic place.

Lastly, please also use this func in `cmd/kubeadm/app/util/apiclient/dryrunclient.go`

File: cmd/kubeadm/app/util/marshal.go (hunk 4 above)
Comment (medinatiger) [reply cmd/kubeadm/app/util/marshal.go]: What's special requirement here? Why is the comments considered not valid?

**Review by medinatiger — COMMENTED** [review]
PTAL

File: cmd/kubeadm/app/util/marshal.go (hunk 4 above)
Comment (ncdc) [reply cmd/kubeadm/app/util/marshal.go]: @medinatiger it needs to follow godoc conventions, where the first word matches the function name (`MarshalToYaml marshals an object into yaml.`).

Comment (medinatiger) [reply cmd/kubeadm/app/util/marshal.go]: @ncdc, thanks! Updated.

**Pull request merged**